libraries/extractor/src/main/java/androidx/media3/extractor/ts/DtsUhdReader.java

```diff
@@ -0,0 +1,242 @@
+    return bytesRead == targetLength;
+  }
+
+  /**
+   * Locates the next SYNC value in the buffer, advancing the position to the byte that immediately
+   * follows it. If SYNC was not located, the position is advanced to the limit.
+   *
+   * @param pesBuffer The buffer whose position should be advanced.
+   * @return Whether SYNC was found.
+   */
+  private boolean skipToNextSync(ParsableByteArray pesBuffer) {
+    while (pesBuffer.bytesLeft() > 0) {
+      syncBytes <<= 8;
+      syncBytes |= pesBuffer.readUnsignedByte();
+      isFtocSync = DtsUtil.isUhdFtocSyncWord(syncBytes);
+      if (isFtocSync || DtsUtil.isUhdFtocNonSyncWord(syncBytes)) {
+        byte[] headerData = headerScratchBytes.getData();
+        headerData[0] = (byte) ((syncBytes >> 24) & 0xFF);
+        headerData[1] = (byte) ((syncBytes >> 16) & 0xFF);
+        headerData[2] = (byte) ((syncBytes >> 8) & 0xFF);
+        headerData[3] = (byte) (syncBytes & 0xFF);
+        bytesRead = 4;
+        syncBytes = 0;
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /** Find the actual size of DTS UHD frame header. */
+  private void findHeaderSize() {
+    byte[] frameData = new byte[FTOC_HEADER_SIZE_MIN];
+    headerScratchBytes.setPosition(0);
+    headerScratchBytes.readBytes(frameData, 0, FTOC_HEADER_SIZE_MIN);
+    headerSizeToRead = DtsUtil.parseDtsUhdHeaderSize(frameData);
+    if (headerSizeToRead < bytesRead) { // Already read more data than the actual header size
+      headerSizeToRead = bytesRead; // Setting target read length equal to bytesRead
+    }
+  }
+
+  /** Parses the sample header. */
+  @RequiresNonNull("output")
+  private void parseHeader() {
+    byte[] frameData = headerScratchBytes.getData();
+    DtsUtil.DtsFormatInfo formatInfo;
+    try {
+      formatInfo = DtsUtil.parseDtsUhdFormat(frameData);
+    } catch (ParserException e) {
+      throw new RuntimeException(e);
+    }
+    if (isFtocSync) { //Format updates will happen only in FTOC sync frames.
+      if (format == null
+          || formatInfo.channelCount != format.channelCount
+          || formatInfo.sampleRate != format.sampleRate
+          || !Util.areEqual(formatInfo.mimeType, format.sampleMimeType)) {
+        format = new Format.Builder()
+            .setId(formatId)
+            .setSampleMimeType(formatInfo.mimeType)
+            .setChannelCount(formatInfo.channelCount)
+            .setSampleRate(formatInfo.sampleRate)
+            .setDrmInitData(null)
+            .setLanguage(language)
+            .build();
+        output.format(format);
+      }
+      sampleCount = formatInfo.sampleCount; // Update the sample count only in FTOC Sync frame
+    }
+    sampleSize = formatInfo.frameSize;
+
+    // In this class a sample is an access unit (frame in DTS), but the format's sample rate
+    // specifies the number of PCM audio samples per second.
+    sampleDurationUs = (int)(C.MICROS_PER_SECOND * sampleCount / format.sampleRate);
```

Comment:
Thanks for pointing out this potential issue.

The sample rate information in the 'formatInfo' is valid only in the FTOC sync frames. For FTOC non-sync frames, we have to use the stream information reported in the last FTOC sync frame.

I will add some handling in the code to avoid any possible NullPointerException.

Reply:
@rohitjoins I have added a fix. Could you review it?

Reply:
Thanks :)